modules/dag/src/main/scala/org/tessellation/dag/l1/DAGBlockValidator.scala

```diff
@@ -0,0 +1,178 @@
+
+import eu.timepit.refined.auto.autoInfer
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import io.estatico.newtype.ops.toCoercibleIdOps
+
+import DAGBlockValidator.takeConsecutiveTransactions
+
+class DAGBlockValidator[F[_]: Async: KryoSerializer: SecurityProvider](
+  transactionStorage: TransactionStorage[F],
+  blockStorage: BlockStorage[F],
+  addressStorage: AddressStorage[F],
+  transactionValidator: TransactionValidator[F]
+) extends BlockValidator[F, DAGBlock] {
+
+  override def accept(signedBlock: Signed[DAGBlock]): F[Unit] =
+    for {
+      _ <- validateBlockSignatures(signedBlock).ifM(
+        Async[F].unit,
+        Async[F].raiseError[Unit](new Throwable("Invalid block signatures!"))
+      )
+      _ <- validateBlock(signedBlock.value).ifM(
+        Async[F].unit,
+        Async[F].raiseError[Unit](new Throwable("Invalid block data!"))
+      )
+      hashedBlock <- signedBlock.hashWithSignatureCheck.flatMap(_.liftTo[F])
+
+      hashedTransactions <- signedBlock.value.transactions.toList
+        .traverse(_.hashWithSignatureCheck.flatMap(_.liftTo[F]))
+
+      _ <- acceptTransactions(hashedTransactions)
+      _ <- blockStorage.acceptBlock(hashedBlock)
+      // TODO: could happen outside of acceptance flow???
+      _ <- blockStorage.handleTipsUpdate(hashedBlock)
+    } yield ()
+
+  def validateBlockSignatures(signedBlock: Signed[DAGBlock]): F[Boolean] =
+    for {
+      areSignaturesValid <- signedBlock.hasValidSignature
+      areAtLeastThreeSignatures = signedBlock.proofs.map(_.id).toList.toSet.size >= 3 //At least 3 distinct signers
+    } yield areSignaturesValid && areAtLeastThreeSignatures
+
+  def validateBlock(block: DAGBlock): F[Boolean] =
+    for {
+      areParentsAccepted <- areParentsAccepted(block)
+      areTxsValid <- areTransactionsValid(block.transactions.toList)
+    } yield areParentsAccepted && areTxsValid
+
+  private def areParentsAccepted(block: DAGBlock): F[Boolean] =
+    blockStorage.areParentsAccepted(block)
+
+  // TODO: for now I think last transaction stored is enough, I'm not storing all the transactions in transaction
+  //       service as it was the case in constellation - all in all, we will store blocks and redownload based on them
+  //       and anyway the check for transactions forming a correct chain should catch this case of accepted or stale transactions trying to pass validation
+//  def containsAcceptedOrStaleTx(cbd: DAGBlockData): F[List[String]] = {
+//    val containsAccepted = cbd.transactions.toList.map { t =>
+//
+//      transactionStorage.getLastAcceptedTransactionRef(t.baseHash).map {
+//        //b => (t.hash, b)
+//        case Some(tr) =>
+//        case None =>
+//      }
+//    }.sequence[F, (String, Boolean)]
+//      .map(l => l.collect { case (h, true) => h })
+//
+//    containsAccepted
+//  }
+
+  private def areTransactionsValid(signedTransactions: Seq[Signed[Transaction]]): F[Boolean] =
```

Comment:
We are using a `Set` of transactions in a block, but I don't think we need to restrict validators to `Set` as most of them would actually convert it to a `Seq` at some point to perform the validations. And if we would actually get a `Seq` of transactions from somewhere we can validate it and fail on duplicates whereas if someone goes from `Seq` to `Set` it will deduplicate during the conversion and it could pass validation when it shouldn't based on the original data. WDYT?